### PR TITLE
Add UI tests of image optimization setting

### PR DESCRIPTION
**Pull request by fluiddot**
This PR is a follow-up of https://github.com/wordpress-mobile/WordPress-iOS/pull/21981 and adds UI tests of the new App Setting to optimize images when uploading.

**To test:**
N/A

## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/UITestsFoundation/XCTestCase+Utils.swift [hunk 1]
```diff
@@ -27,7 +27,7 @@ public extension XCTestCase {
 
         appToRemove.firstMatch.press(forDuration: 1)
         waitAndTap(Apps.springboard.buttons["Remove App"])
-        waitAndTap(Apps.springboard.alerts.buttons["Delete App"])
+        waitForExistenceAndTap(Apps.springboard.alerts.buttons["Delete App"])
```
Comment (fluiddot) [inline WordPress/UITestsFoundation/XCTestCase+Utils.swift:30]: I noticed that the `Delete app` button wasn't tapped when running the UI tests locally. For this reason, I decided to change this action to explicitly wait for the element's existence.

**Comment by twstokes** [comment]
Nice @fluiddot! Thanks for going the extra mile to add these.

**Comment by pachlava** [comment]
@fluiddot 👋 

Just leaving a note that I'm looking into a PR — will leave some feedback today a bit later. Thank you for adding the tests! 👍

File: WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift [hunk 2]
```diff
@@ -96,6 +96,14 @@ public class BlockEditorScreen: ScreenObject {
         $0.staticTexts["You have unsaved changes."]
     }
 
+    private let turnOnImageOptimizationGetter: (XCUIApplication) -> XCUIElement = {
+        $0.buttons["Yes, leave on"]
+    }
+
+    private let turnOffImageOptimizationGetter: (XCUIApplication) -> XCUIElement = {
+        $0.buttons["No, turn off"]
+    }
```
Comment (pachlava) [inline WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift]: For getters, we follow the naming pattern where the element type is included:

```suggestion
    private let turnOnImageOptimizationButtonGetter: (XCUIApplication) -> XCUIElement = {
        $0.buttons["Yes, leave on"]
    }

    private let turnOffImageOptimizationButtonGetter: (XCUIApplication) -> XCUIElement = {
        $0.buttons["No, turn off"]
    }
```

File: WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift [hunk 3]
```diff
@@ -119,6 +127,8 @@ public class BlockEditorScreen: ScreenObject {
     var switchToHTMLModeButton: XCUIElement { switchToHTMLModeButtonGetter(app) }
     var undoButton: XCUIElement { undoButtonGetter(app) }
     var unsavedChangesLabel: XCUIElement { unsavedChangesLabelGetter(app) }
+    var turnOnImageOptimization: XCUIElement { turnOnImageOptimizationGetter(app) }
+    var turnOffImageOptimization: XCUIElement { turnOffImageOptimizationGetter(app) }
```
Comment (pachlava) [inline WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift]: Same sentiment here:

```suggestion
    var turnOnImageOptimizationButton: XCUIElement { turnOnImageOptimizationButtonGetter(app) }
    var turnOffImageOptimizationButton: XCUIElement { turnOffImageOptimizationButtonGetter(app) }
```

File: WordPress/UITests/Tests/AppSettingsTests.swift [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    func testImageOptimizationIsTurnedOnEditor() throws {
+        try TabNavComponent().goToBlockEditorScreen().addImage().chooseOptimizeImages(option: true).closeEditor()
+        try TabNavComponent().goToMeScreen().goToAppSettings().verifyImageOptimizationSwitch(enabled: true)
+    }
+
+    func testImageOptimizationIsTurnedOffEditor() throws {
+        try TabNavComponent().goToBlockEditorScreen().addImage().chooseOptimizeImages(option: false).closeEditor()
+        try TabNavComponent().goToMeScreen().goToAppSettings().verifyImageOptimizationSwitch(enabled: false)
+    }
```
Comment (pachlava) [inline WordPress/UITests/Tests/AppSettingsTests.swift]: Generally, we keep the method chaining start on a new line for each call, to make debugging of separate steps possible:

```suggestion
    func testImageOptimizationEnabledByDefault() throws {
        try TabNavComponent()
            .goToMeScreen()
            .goToAppSettings()
            .verifyImageOptimizationSwitch(enabled: true)
    }

    func testImageOptimizationIsTurnedOnEditor() throws {
        try TabNavComponent()
            .goToBlockEditorScreen()
            .addImage()
            .chooseOptimizeImages(option: true)
            .closeEditor()

        try TabNavComponent()
            .goToMeScreen()
            .goToAppSettings()
            .verifyImageOptimizationSwitch(enabled: true)
    }

    func testImageOptimizationIsTurnedOffEditor() throws {
        try TabNavComponent()
            .goToBlockEditorScreen()
            .addImage()
            .chooseOptimizeImages(option: false)
            .closeEditor()

        try TabNavComponent()
            .goToMeScreen()
            .goToAppSettings()
            .verifyImageOptimizationSwitch(enabled: false)
    }
```

File: WordPress/UITests/Tests/AppSettingsTests.swift [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import UITestsFoundation
+import XCTest
+
+final class AppSettingsTests: XCTestCase {
+
+    @MainActor
+    override func setUpWithError() throws {
+        try super.setUpWithError()
+        setUpTestSuite(removeBeforeLaunching: true)
```
Comment (pachlava) [inline WordPress/UITests/Tests/AppSettingsTests.swift]: At first, I was confused why this file uses `removeBeforeLaunching`, but then I realized that this is to ensure that Image Optimization prompt is shown for both `testImageOptimizationIsTurnedOnEditor` and `testImageOptimizationIsTurnedOffEditor`, and that tests in this file are independent/isolated.

We might need to keep this in mind, though, if we add more tests into this file: for other tests deleting the app might not be necessary, but it takes quite a lot of time.

**Review by pachlava — APPROVED** [review]
@fluiddot 👋 

Thank you so much for adding the UI tests, and extending the Screens to support "App Settings" screen ⭐ 

I executed the tests locally for iPhone with no issues 👍 , and left two minor naming / formatting comments, they're not blocking. :shipit: 

P.S. One thing to discuss (@jostnes and @tiagomar 👋): since we try to have mainly critical flows automated, should we consider splitting the tests schemes to `Smoke` and `Full`, to keep the execution time under 30 minutes for `Smoke`? With Xcode 15 incoming, this might become problematic if we keep on running all tests on every push.

File: WordPress/UITests/Tests/AppSettingsTests.swift (hunk 5 above)
Comment (fluiddot) [reply WordPress/UITests/Tests/AppSettingsTests.swift]: > At first, I was confused why this file uses `removeBeforeLaunching`, but then I realized that this is to ensure that Image Optimization prompt is shown for both `testImageOptimizationIsTurnedOnEditor` and `testImageOptimizationIsTurnedOffEditor`, and that tests in this file are independent/isolated.

Yeah, I should have added a note here explaining why it's necessary to remove it before launching. I'll add an inline comment as a follow-up.

> We might need to keep this in mind, though, if we add more tests into this file: for other tests deleting the app might not be necessary, but it takes quite a lot of time.

I agree. This approach is specific to the tests I added. In the future, it's likely to be unnecessary for other tests related to the App Settings screen. I wonder if there's an option to enable `removeBeforeLaunching` at the test level. If not, we can consider renaming the test file to `OptimizeImagesSettingTests` to convey that the file should only include test cases related to that logic.

Comment (pachlava) [reply WordPress/UITests/Tests/AppSettingsTests.swift]: > I wonder if there's an option to enable `removeBeforeLaunching` at the test level

Maybe not the nicest way, but this should do the trick (not tested):

```swift
let testsRequiringAppDeletion = [
        "testImageOptimizationEnabledByDefault",
        "testImageOptimizationIsTurnedOnEditor",
        "testImageOptimizationIsTurnedOffEditor"
    ]

let removeBeforeLaunch = testsRequiringAppDeletion.contains(where: { testName in self.name.contains(testName)})

try super.setUpWithError()
setUpTestSuite(removeBeforeLaunching: removeBeforeLaunch)
```

File: WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift (hunk 2 above)
Comment (fluiddot) [reply WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift]: Thanks @pachlava for the suggestion, I've applied this in https://github.com/wordpress-mobile/WordPress-iOS/pull/22095/commits/88e54245ae1d1eafc80e92ce4ec51a4ae2820140.

File: WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift (hunk 3 above)
Comment (fluiddot) [reply WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift]: Suggestion applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/22095/commits/88e54245ae1d1eafc80e92ce4ec51a4ae2820140.

File: WordPress/UITests/Tests/AppSettingsTests.swift (hunk 4 above)
Comment (fluiddot) [reply WordPress/UITests/Tests/AppSettingsTests.swift]: Ah good point, thanks for the call out, I applied this suggestion in https://github.com/wordpress-mobile/WordPress-iOS/pull/22095/commits/ac3d4d4e664be4a6c19c2bd06403f65dedbc1ccf.

File: WordPress/UITests/Tests/AppSettingsTests.swift (hunk 5 above)
Comment (fluiddot) [reply WordPress/UITests/Tests/AppSettingsTests.swift]: I tried your approach and it worked nice, thanks for sharing it. I haven't found another approach to use a different setup depending on the test case, so I think we can apply this workaround. Applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/22095/commits/00e4d46e825778b4d84d968fad1d6a36fab1a7a3.

**Pull request merged**